subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/report/HtmlDependencyVerificationReportRendererTest.groovy

```diff
@@ -0,0 +1,327 @@
+        return wrap(repo, new SignatureVerificationFailure(dummyFile, dummyFileSig, errors, new DummyKeyService()))
+    }
+
+    private static SignatureVerificationFailure.SignatureError signatureError(SignatureVerificationFailure.FailureKind kind = MISSING_KEY) {
+        return new SignatureVerificationFailure.SignatureError(null, kind)
+    }
+
+    private static RepositoryAwareVerificationFailure wrap(String repo, VerificationFailure vf) {
+        return new RepositoryAwareVerificationFailure(vf, repo)
+    }
+
+    private static ModuleComponentArtifactIdentifier artifact(String group = "org", String name = "foo", String version = "1.0", String fileName = "foo-1.0.jar") {
+        return new ModuleComponentFileArtifactIdentifier(
+            DefaultModuleComponentIdentifier.newId(DefaultModuleIdentifier.newId(group, name), version),
+            fileName
+        )
+    }
+
+    private void generateReport() {
+        currentReportFile = renderer.writeReport()
+        currentReportDir = currentReportFile.parentFile
+        def parser = new SAXParser()
+        def slurper = new XmlSlurper(parser)
+        report = slurper.parse(currentReportFile)
+    }
+
+    private boolean bodyContains(String text) {
+        def node = report.BODY.depthFirst().find { norm(it).contains(text) }
+        node != null // this is just so that we can put a breakpoint for debugging
+    }
+
+    private boolean bodyContainsExact(String text) {
+        def node = report.BODY.depthFirst().find { norm(it) == text }
+        node != null // this is just so that we can put a breakpoint for debugging
+    }
+
+    private static String norm(NodeChild node) {
+        norm(node.text())
+    }
+
+    private static String norm(String s) {
+        // replaces non-breaking space with space, makes testing easier
+        s.replace(' ', ' ').trim()
+    }
+
+    private List<ReportedError> errorsFor(String section) {
+        NodeChild handle = report.BODY.depthFirst().find { it.@class == "uk-accordion-title" && it.text().startsWith(section) }
+        handle.parent().DIV.TABLE.TBODY.TR.collect {
+            new ReportedError(it)
+        }
+    }
+
+    private static class ReportedError {
+        final NodeChild row
+        final String module
+        final String artifact
+        final String problem
+        final boolean hasSignature
+
+        ReportedError(NodeChild row) {
+            this.row = row
+            module = norm(row.TD[0])
+            def artifactText = norm(row.TD[1])
+            artifact = artifactText - ' (.asc)'
+            hasSignature = artifactText.contains(" (.asc)")
+            problem = norm(row.TD[2])
+        }
+
+        String getArtifactTooltip() {
+            (row.TD[1].DIV.@'uk-tooltip').text() - 'title: '
+        }
+    }
+
+    private static class DummyKeyService implements PublicKeyService {
```

Comment:
This is only because of reporting of errors and dependency injection, where we want to extract some information from the key. Given this code actually refactors how the errors are generated, we could definitely remove the `explainTo` method which requires it.